app/controllers/cases/notes_controller.rb

```diff
@@ -0,0 +1,51 @@
+module Cases
+  class NotesController < ApplicationController
+    before_action :set_case, :set_date_of_birth
+
+    def create
+      begin
+        @case.state_machine.add_note_to_case!(
+          acting_user: current_user,
+          acting_team: case_team,
+          message: params[:case][:message_text]
+        )
+      rescue ActiveRecord::RecordInvalid => err
+        if err.record.errors.include?(:message)
+          flash[:case_errors] = { message_text: err.record.errors[:message] }
+        end
+      ensure
+        redirect_to case_path(@case, anchor: 'case-history')
+      end
+    end
```

Comment:
Existing pattern is to put the `create` block of code into a dedicated Service.

Reply:
Also should this controller be in the `/controllers/cases` namespace so that the URL is `/cases/:id/notes` ?

Reply:
Have I understood correctly that `Notes` and `Case History` are literally the same thing but just have different outputs? If so can `messages_controller` not be repurposed? Or if that is undesirable a `module` called `Notes` or similar. `messages_controller` is already inside `/controllers/cases/` with appropriate routes

Reply:
This is basically a copy of messages_controller and the way it works is the same, "Existing pattern is to put the create block of code into a dedicated Service" isn't the case for messages controller. They need to redirect differently and need to save with a different event, so I think having a separate route and controller is fine, actually, and better than adding complexity to messages controller

Reply:
Oh, and although the two things, notes and messages, use the same json property name, I think they're different concepts and deserve their own controller. A key point being you can in theory use both alongside each other.

Reply:
Agree with them being different controllers. Weary we are pretending they are different through the actual domain mechanism as well. Just thinking about this from a testing perspective, as a result of the duplicated code the actual persistence mechanism is tested twice. And with two different implementations the risk of diverging inadvertently increases IMO. 

What I had in mind was something akin to the following for example:

```
+controllers/concerns/notable.rb+

module Notable
  def add(event_name:, on_success:)
    begin
      @case = Case::Base.find params[:case_id]
      @case.state_machine.send(event_name,
        acting_user: current_user,
        acting_team: case_team,
        message: params[:case][:message_text],
        controller: self.class
      )
    rescue ActiveRecord::RecordInvalid => err
      if err.record.errors.include?(:message)
        flash[:case_errors] = { message_text: err.record.errors[:message] }
      end
    ensure
      redirect_to on_success
    end    
  end

  private

  # Need to sort with manager first so that if we are both manager and something else, we don't
  # try to execute the action with our lower authority (which might fail)
  def case_team
    User.sort_teams_by_roles(case_teams).first
  end  

  def case_teams
    if current_user.teams_for_case(@case).any?
      current_user.teams_for_case(@case)
    else
      current_user.teams
    end
  end  
end

+cases/notes_controller.rb+

module Cases
  class NotesController < ApplicationController
    include Notable
    include GovUKDateFixes

    # Move this before_action into GovUKDateFixes ?
    before_action :set_date_of_birth

    def create
      add(
        event_name: 'add_note_to_case!',
        on_success:  case_path(@case, anchor: 'case-history')
      )
    end
  end
end

+cases/messages_controller.rb+
module Cases
  class MessagesController < ApplicationController
    include Notable

    def create
      add(
        event_name: 'add_message_to_case!',
        on_success:  case_path(@case, anchor: 'messages-section')
      )
    end
  end
end

+models/case_transition.rb+
# Addition of 'controller' property to allow debugging/retrieval
  jsonb_accessor :metadata,
                 message:                    :text,
                 filenames:                  [:string, array: true, default: []],
                 final_deadline:             :date,
                 linked_case_id:             :integer,
                 original_final_deadline:    :date,
                 controller:                 :string
```